src/web/arr/trove/chart.arr

```diff
@@ -122,6 +285,367 @@ y-max-method = method(self, y-max :: Number):
+  neg-step = step-types.filter({(n): n >= num-abs(neg-sci.coeff / 9)}).get(0) * num-expt(10, neg-sci.exponent)
+  step = num-max(pos-step, neg-step)
+  step-sci = scientific-b10(step)
+
+  # Use step distance to calculate Axis Properties
+  name-tick = 
+    {(n): 
+      ask:
+      | (step-sci.coeff == 2.5) and (step-sci.exponent <= 0) then: 
+        pointer(num-to-string-digits(n, 2 - step-sci.exponent), n)
+      | step-sci.exponent < 0 then: 
+        pointer(num-to-string-digits(n, 1 - step-sci.exponent), n)
+      | otherwise: 
+        pointer(num-to-string(n), n)
+      end}
+
+  axisTop = num-max(0, step * num-ceiling(pos-bar-height / step))
+  axisBottom = num-min(0, step * num-floor(neg-bar-height / step))
+  pos-ticks = map(name-tick, range-by(0, axisTop + step, step))
+  neg-ticks = map(name-tick, range-by(0, axisBottom - step, -1 * step))
+
+  self.constr()(
+    self.obj.{axisdata: some(axis-data(axisTop, axisBottom, distinct(pos-ticks + neg-ticks)))}
+    )
+end
+
+format-axis-data-method = method(self, format-func :: (Number -> String)):
+  cases (Option) self.obj!axisdata: 
+    | none => 
+      raise("Should never have reached this point. Yell at John for not setting up the axis properties somewhere where he should have and please report this as a bug")
+    | some(ad) => 
+      new-ticks = map({(p): pointer(format-func(p.value), p.value)}, ad.ticks)
+      self.constr()(self.obj.{axisdata: some(axis-data(ad.axisTop, ad.axisBottom, new-ticks))})
+  end
+end
+
+scale-method = method(self, scale-fun :: (Number -> Number)): 
+  exact-sf = {(n): n ^ scale-fun ^ num-to-rational}
+  list-of-rows = self.obj!tab ^ raw-array-to-list
+  scale-row = {(row): [raw-array: raw-array-get(row, 0), raw-array-get(row, 1) ^ exact-sf]}
+  scaled-tab = map(scale-row, list-of-rows) ^ builtins.raw-array-from-list
+  scaled-self = self.constr()(self.obj.{tab: scaled-tab})
+  scaled-values = map({(row): raw-array-get(row, 1) ^ exact-sf}, list-of-rows)
+  {max-positive-height; max-negative-height} = prep-axis(scaled-values)
+
+  scaled-self.make-axis(max-positive-height, max-negative-height)
+end
+
+multi-scale-method = method(self, scale-fun :: (Number -> Number)): 
+  exact-sf = {(n): n ^ scale-fun ^ num-to-rational}
+  list-of-rows = self.obj!tab ^ raw-array-to-list
+  get-values = {(row): raw-array-get(row, 1) ^ raw-array-to-list}
+  scale-row = {(row): [raw-array: raw-array-get(row, 0), map(exact-sf, row ^ get-values) ^ builtins.raw-array-from-list]}
+  scaled-tab = map(scale-row, list-of-rows) ^ builtins.raw-array-from-list
+  scaled-self = self.constr()(self.obj.{tab: scaled-tab})
+  scaled-values = map({(row): map(exact-sf, row ^ get-values)}, list-of-rows)
+  {max-positive-height; max-negative-height} = 
+    multi-prep-axis(scaled-self.obj!is-stacked, scaled-values)
+
+  scaled-self.make-axis(max-positive-height, max-negative-height)
+end
+
+stacking-type-method = method(self, stack-type :: String): 
```

Comment:
Does it make more sense to define a new data definition, rather than to use a string, for `stack-type`?

Reply:
I think it makes some sense. 

The main difference between the two series is that 
- Bar-chart-series has its data represented by a List<Number> which contains the heights of the bars. 
- Multi-bar-chart-series has its data represented by a List<List<Number>> which contains the heights of the bars separated into components. 

The reason this is important is that many of the methods shown above must be implemented differently in each case. For example: 
- Colors is defined to color each bar in the bar-chart-series case and each component in the multi-bar-chart-series 
- Sort-by-data only makes sense in the context of a multi-bar-chart-series, where you might want to only sort by certain components 
- Calculating the axis depends on the highest bar displayed and the method needed to find the changes for each data-series. 
- Scale also has to be mapped differently based on each data-series 

Also stacking type would have to be able to change between the List<List<Number>> (multi-bar-chart-series) data and List<Number> (bar-chart-series) data. 
While we can switch from multi-bar-chart-series to bar-chart-series naturally,  It doesn’t seem feasible to go from bar-chart-series to multi-bar-chart-series naturally.

Reply:
Hmm. I'm not sure if I understand your comment. What I meant is that we could define:

```
data StackType:
  | absolute
  | relative
  | percent
  | none
end
```

and use these instead of string.

Reply:
Oh I'm sorry I misunderstood what you saying. I think that forcing the users to create an new stacktype instance every time they want to change stacking types might be a bit too cumbersome and maybe too complicated/uncommon for anyone just starting out learning pyret. Thus personally I believe using strings would be better if only because they would be more familiar.

Reply:
> I think that forcing the users to create an new stacktype instance every time they want to change stacking types might be a bit too cumbersome

I don't think it's any more cumbersome than creating stuff like `mode-solid` in the `image-typed` library.

> maybe too complicated/uncommon for anyone just starting out learning pyret

My goal is to provide a good interface for people to use. If some audiences need a very simple interface, wrapper libraries could be created for that purpose, and this is already the case for the chart library for Bootstrap.

`Option` looks pretty complicated compared to `null`. One can argue that people learning programming would understand `null` better, and that might even be true. Still, there is value in learning the right way to do things.

Reply:
Ah I see that makes sense, it's actually not cumbersome at all. Sorry, I must've gotten a bit confused from jumping around different coding languages, for some reason I thought you would have to make a new stacking type instance every time you called it like with classes. I actually have the opposite opinion now in that the strings might be unnecessary. Or should we do it the same was as Fillmode where we support both. If so should we split it into an typed and untyped library?

Reply:
I think the fact that `Fillmode` is different in the typed and untyped library is mostly because the typed one is better and is what we want, but we can't remove the untyped one due to backward incompatibility. (the untyped interface has been around for a very long time)

One of the reasons why I am pushing to have `StackType` now is that if we wait longer and people start to use the current interface, then we won't be able to change the interface anymore due to the backward incompatibility constraint.

Reply:
It's not quite that simple, either. In principle I agree with you, but in practice, using enumerated values doesn't necessarily make things simpler for teaching. Now you have a curricular dependency on explaining what these magic variable names mean, whereas the string-based API only requires teaching about strings.

I think possibly the right thing to do here is make *this* API be an enumeration, and ask @schanzer to wrap the API in a stringier version for use in bootstrap teachpack libraries. @schanzer and @jpolitz , agreed?